### ISSUE TITLE
[💡SUG] 可以给一个标准的LightGCN的复现配置吗？

### ISSUE THREAD

**Issue by h12345jack** [RUCAIBox/RecBole#1070]
你好，我最近在复现LightGCN，[相关的代码库](https://github.com/gusye1234/LightGCN-PyTorch)是可以复现文章的结果的。
但是如果使用recbole, 和文章有一定的差距，我看[之前的issue也有类似的情况](https://github.com/RUCAIBox/RecBole/issues/1020)，
可以帮我看看原因吗？ 

首先数据集:

[yelp2018.zip](https://github.com/RUCAIBox/RecBole/files/7641657/yelp2018.zip)

这是我处理的文章中提到的数据成Recbole可以使用的格式。
配置:

```
# run it in order to validate LightGCN results
seed: 2020
# model config
embedding_size: 64
n_layers: 2
reg_weight: 1e-05

# dataset config
field_separator: "," #指定数据集field的分隔符
seq_separator: " " #指定数据集中token_seq或者float_seq域里的分隔符
USER_ID_FIELD: userId #指定用户id域
ITEM_ID_FIELD: itemId #指定物品id域
RATING_FIELD: rating #指定打分rating域
TIME_FIELD: timestamp #指定时间域
NEG_PREFIX: neg_ #指定负采样前缀
LABEL_FIELD: label #指定标签域
#指定从什么文件里读什么列，这里就是从ml-1m.inter里面读取user_id, item_id, rating, timestamp这四列
load_col:
  inter: [userId, itemId]

# training settings
epochs: 1000 #训练的最大轮数
train_batch_size: 2048 #训练的batch_size
learner: adam #使用的pytorch内置优化器
learning_rate: 0.001 #学习率
training_neg_sample_num: 1 #负采样数目
eval_step: 2 #每次训练后做evalaution的次数
neg_sampling:
  uniform: 1
stopping_step: 5 #控制训练收敛的步骤数，在该步骤数内若选取的评测标准没有什么变化，就可以提前停止了

# evalution settings
metrics: ["Precision","Recall","NDCG"] #评测标准
topk: [10, 15, 20] #评测标准使用topk，设置成10评测标准就是["Recall@10", "MRR@10", "NDCG@10", "Hit@10", "Precision@10"]
valid_metric: Recall@20 #选取哪个评测标准作为作为提前停止训练的标准
eval_batch_size: 4096 #评测的batch_size
loss_decimal_place: 4
metric_decimal_place: 4


# gpu settings
gpu_id: 1


# others for my task
benchmark_filename: ['train', 'test', 'test"]

data_path: 'datas/'
show_progress: True

eval_args:
  split:
```

跑出来的结果:

```
02 Dec 19:40    INFO  best valid : {'precision@10': 0.0299, 'precision@15': 0.0277, 'precision@20': 0.0261, 'recall@10': 0.0331, 'recall@15': 0.046, 'recall@20': 0.058, 'ndcg@10': 0.0382, 'ndcg@15': 0.0427, 'ndcg@20': 0.0473}
02 Dec 19:40    INFO  test result: {'precision@10': 0.0299, 'precision@15': 0.0277, 'precision@20': 0.0261, 'recall@10': 0.0331, 'recall@15': 0.046, 'recall@20': 0.058, 'ndcg@10': 0.0382, 'ndcg@15': 0.0427, 'ndcg@20': 0.0473}

```

|         |. Recall@20     |    ndcg@20   | 
| ------- | ------------- | -------------|
| Recbole结果 | 0.058 |     0.0473  |
| 原论文 | 0.0622 |   0.0504 | 

差距是8% , (0.0622 - 0.058)  / 0.058 = 8%, 这还是没有划分val的结果，如果 7:1:2进行实验，效果还要差一些，请问可以分析一下原因吗？

**Comment by Sherry-XLL** [comment]
@h12345jack 您好，#1020 的回复已经说明了原因，由于 RecBole 是一个统一、全面、高效的推荐系统代码库，数据处理和代码框架不是为了复现某个特定的模型，而是能让不同的模型在统一的评测体系下能有一个相对的结果。

如果您的目的是复现论文的数据，那么您需要保证所有细节都与论文的处理一致，比如：

- 在原论文中有提到 `We optimize LightGCN with Adam and use the default learning rate of 0.001 and default mini-batch size of 1024 (on Amazon-Book, we increase the mini-batch size to 2048 for speed).` 因此 Yelp2018 数据集的 `train_batch_size` 应该设置为 1024。
- 另外，`The early stopping and validation strategies are the same as NGCF.`，在 NGCF 中 `premature stopping if recall@20 on the validation data does not increase for 50 successive epochs.`，即 `stopping_step` 应该设置为 50。

LightGCN 的收敛速度相对较慢，`stopping_step` 设置为 5 可能会使得模型过早停止，也许这是您的结果欠佳的原因之一。值得一提的是，即使所有配置都和论文完全一致，也无法得到完全相同的结果，超参数微调、环境差异、Pytorch 版本、数据处理细节等原因都会使得结果在一定范围内浮动，为了避免争议，RecBole 并不会提供各个模型复现的具体数值，配置细节还需使用者自己调整。

**Comment by h12345jack** [comment]
@Sherry-XLL 非常理解“结果在一定范围内浮动”。
不过由于需要评估不同的模型的性能，如果没有一个“可复现“的性能，不太利于进一步模型的设计。
很多的论文中都没有重新进行实验，而是copy from paper，前提都是在同一个数据集和划分中，但是如果recbole不能很好的复现论文中的结果，可能无法 “能让不同的模型在统一的评测体系下能有一个相对的结果”。

类似的框架中PyG在[原始论文](https://arxiv.org/abs/1903.02428)中也给出了其复现的GCN和GAT的结果。

**Comment by h12345jack** [comment]
@Sherry-XLL  在提升了stopping_step=50, batch_size= 1024后，实验结果0.059 （Recall) 和 0.0481 (NDCG)。

**Comment by 2017pxy** [comment]
@h12345jack 你好，原论文给的结果是用TensorFlow的代码跑的，TensorFlow和PyTorch的底层还是存在一定差异的，所以存在差异是很正常的；我看了下给出的LightGCN的PyTorch源码，发现他们在layer = 2 时的 recall是 0.05988， ndcg是0.04956，这和你给出的这个结果相比：

> 在提升了stopping_step=50, batch_size= 1024后，实验结果0.059 （Recall) 和 0.0481 (NDCG)

我认为差异已经是在一个可以容忍的范围了。

此外，我认为，只要是两种框架下的实现方式，即便是超参和实验配置完全一样，结果也有很大概率是不同的（会在一个可接受的范围内波动）。因为两种框架的数据处理，batch 划分方式，随机采样等可能都会有差异，这种差异大多源于不同工程实现的方式，而不是代码开发的错误。因此我认为，不能因为和原文（或者源码）结果无法一一对应，就证明我们的代码实现是不可靠的。

我们做伯乐的初衷，就是为了有一个相对公平的平台来比较不同模型，但考虑到不同模型之间存在着差异，为了建立一个统一的框架，我们需要在各个环节做权衡和调整，不能完全兼顾某一种模型，所以我们的代码开发相比于原文代码，也许会在数据处理等部分有所不同。因此，我们在做模型开发的测试时，标准不仅仅是复现原文结果的绝对值，我们更侧重于比较同一个配置下，模型之间的相对排名和差距。我们现在可以保证的是：我们的模型在相同配置下，是可以跑出一个基本符合先前文献实验结果的排序的，同时模型的结果和原文不会存在过大差异。

当然，每个人对差异的理解和容忍度是不同的，如果你依然觉得我们的实现和原文出入过大，你可以使用原文的标准代码进行实验；我们也欢迎你指出我们代码中的问题或者bug。谢谢！

**Comment by h12345jack** [comment]
在修改了layer= 3后，结果也没有优于layer=2，并且距离文章的结果差距还比较大。
```
{'precision@10': 0.0311, 'precision@15': 0.0285, 'precision@20': 0.0268, 'recall@10': 0.0346, 'recall@15': 0.0475, 'recall@20': 0.0595, 'ndcg@10': 0.0396, 'ndcg@15': 0.0441, 'ndcg@20': 0.0487}
```
我再探索一下差距的原因（也许是初始化导致的）。

**Comment by h12345jack** [comment]
@Sherry-XLL @2017pxy , 你们好，
在探索了一段时间后，我似乎找到了原因。

无论是[Pytorch版本](https://github.com/gusye1234/LightGCN-PyTorch/blob/e8a37ba1bb0affb2f93dabeaa23f676182da4164/code/model.py#L194) 还是 (TF版本)(https://github.com/kuandeng/LightGCN/blob/b067ec05f16a24e0e1efe63cd82385f0c469af20/LightGCN.py#L344), 
使用的L2 正则都是without the sqrt, refer to [this](https://www.tensorflow.org/api_docs/python/tf/nn/l2_loss)

因此，需要进行以下的修改，

```
reg_loss = (1/2)*(u_ego_embeddings.norm(2).pow(2) + 
                          pos_ego_embeddings.norm(2).pow(2) + 
                          neg_ego_embeddings.norm(2).pow(2))/float(len(interaction))
```

事实上，在[RegLoss](https://github.com/RUCAIBox/RecBole/blob/eb712510449b53d9459f79178f2870cec84e964e/recbole/model/loss.py#L49) 和 [EmbLoss](https://github.com/RUCAIBox/RecBole/blob/eb712510449b53d9459f79178f2870cec84e964e/recbole/model/loss.py#L67)都可能需要进行修改，在L2正则的情况下。

**Comment by h12345jack** [comment]
@2017pxy @Sherry-XLL 
Any updates?
I think your L2 regularization is not implemented correctly.
Please refer to [this thread](https://discuss.pytorch.org/t/how-does-one-implement-weight-regularization-l1-or-l2-manually-without-optimum/7951/16)

**Comment by 2017pxy** [comment]
@h12345jack 你好，我们已经注意到了这个问题，正在进行修复和测试，感谢反馈

**Comment by Wicknight** [comment]
@h12345jack 您好！请问您具体是如何修改的呢？我在按照上述提议修改EmbLoss后在您提供的配置下（已修改train_batch_size和stopping_step）运行得到如下结果：
`test result: {'precision@10': 0.0306, 'precision@15': 0.0284, 'precision@20': 0.0266, 'recall@10': 0.034, 'recall@15': 0.0473 'recall@20': 0.0591, 'ndcg@10': 0.0391, 'ndcg@15': 0.0438, 'ndcg@20': 0.0483}`
相对应地，在修改EmbLoss前，在配置下运行结果如下：
`test result: {'precision@10': 0.0306, 'precision@15': 0.0282, 'precision@20': 0.0265, 'recall@10': 0.034, 'recall@15': 0.047, 'recall@20': 0.0588, 'ndcg@10': 0.039, 'ndcg@15': 0.0435, 'ndcg@20': 0.048}`
二者的差异非常小，因此想请教下您更加具体的配置以及修改内容以便于我们进行测试，也欢迎您提交pull request来做修改！

**Comment by h12345jack** [comment]
@Wicknight，你好

我跑的是n_layers=3的结果。

```
test result: {'precision@10': 0.0332, 'precision@15': 0.0302, 'precision@20': 0.0284, 'recall@10': 0.0368, 'recall@15': 0.0504, 'recall@20': 0.063, 'ndcg@10': 0.0425, 'ndcg@15': 0.0471, 'ndcg@20': 0.0519}
```



```

# -*- coding: utf-8 -*-
# @Time   : 2020/8/31
# @Author : Changxin Tian
# @Email  : cx.tian@outlook.com

# UPDATE:
# @Time   : 2020/9/16
# @Author : Shanlei Mu
# @Email  : slmu@ruc.edu.cn

r"""
LightGCN
################################################

Reference:
    Xiangnan He et al. "LightGCN: Simplifying and Powering Graph Convolution Network for Recommendation." in SIGIR 2020.

Reference code:
    https://github.com/kuandeng/LightGCN
"""

import numpy as np
import scipy.sparse as sp
import torch
import torch.nn as nn


from recbole.model.abstract_recommender import GeneralRecommender
from recbole.model.init import xavier_uniform_initialization
from recbole.model.loss import BPRLoss
from recbole.utils import InputType


class EmbLoss(nn.Module):
    """ EmbLoss, regularization on embeddings

    """

    def __init__(self, norm=2):
        super(EmbLoss, self).__init__()
        self.norm = norm

    def forward(self, *embeddings):
        emb_loss = torch.zeros(1).to(embeddings[-1].device)
        for embedding in embeddings:
            emb_loss += torch.norm(embedding, p=self.norm).pow(2)
        emb_loss /= embeddings[-1].shape[0]
        return emb_loss


class LightGCN(GeneralRecommender):
    r"""LightGCN is a GCN-based recommender model.

    LightGCN includes only the most essential component in GCN — neighborhood aggregation — for
    collaborative filtering. Specifically, LightGCN learns user and item embeddings by linearly 
    propagating them on the user-item interaction graph, and uses the weighted sum of the embeddings
    learned at all layers as the final embedding.

    We implement the model following the original author with a pairwise training mode.
    """
    input_type = InputType.PAIRWISE

    def __init__(self, config, dataset):
        super(LightGCN, self).__init__(config, dataset)

        # load dataset info
        self.interaction_matrix = dataset.inter_matrix(form='coo').astype(np.float32)

        # load parameters info
        self.latent_dim = config['embedding_size']  # int type:the embedding size of lightGCN
        self.n_layers = config['n_layers']  # int type:the layer num of lightGCN
        self.reg_weight = config['reg_weight']  # float32 type: the weight decay for l2 normalization

        # define layers and loss
        self.user_embedding = torch.nn.Embedding(num_embeddings=self.n_users, embedding_dim=self.latent_dim)
        self.item_embedding = torch.nn.Embedding(num_embeddings=self.n_items, embedding_dim=self.latent_dim)
        self.mf_loss = BPRLoss()
        self.reg_loss = EmbLoss()

        # storage variables for full sort evaluation acceleration
        self.restore_user_e = None
        self.restore_item_e = None

        # generate intermediate data
        self.norm_adj_matrix = self.get_norm_adj_mat().to(self.device)

        # parameters initialization
        # self.apply(kaiming_uniform_initialization)
        
        # use default initialization
        nn.init.normal_(self.user_embedding.weight, std=0.1)
        nn.init.normal_(self.item_embedding.weight, std=0.1)

        self.other_parameter_name = ['restore_user_e', 'restore_item_e']

    def get_norm_adj_mat(self):
        r"""Get the normalized interaction matrix of users and items.

        Construct the square matrix from the training data and normalize it
        using the laplace matrix.

        .. math::
            A_{hat} = D^{-0.5} \times A \times D^{-0.5}

        Returns:
            Sparse tensor of the normalized interaction matrix.
        """
        # build adj matrix
        A = sp.dok_matrix((self.n_users + self.n_items, self.n_users + self.n_items), dtype=np.float32)
        inter_M = self.interaction_matrix
        inter_M_t = self.interaction_matrix.transpose()
        data_dict = dict(zip(zip(inter_M.row, inter_M.col + self.n_users), [1] * inter_M.nnz))
        data_dict.update(dict(zip(zip(inter_M_t.row + self.n_users, inter_M_t.col), [1] * inter_M_t.nnz)))
        A._update(data_dict)

        sumArr = (A > 0).sum(axis=1)
        # add epsilon to avoid divide by zero Warning
        diag = np.array(sumArr.flatten())[0] + 1e-7
        diag = np.power(diag, -0.5)
        D = sp.diags(diag)
        L = D * A * D
        # covert norm_adj matrix to tensor
        L = sp.coo_matrix(L)
        row = L.row
        col = L.col
        i = torch.LongTensor([row, col])
        data = torch.FloatTensor(L.data)
        SparseL = torch.sparse.FloatTensor(i, data, torch.Size(L.shape)).coalesce()
        return SparseL

    def get_ego_embeddings(self):
        r"""Get the embedding of users and items and combine to an embedding matrix.

        Returns:
            Tensor of the embedding matrix. Shape of [n_items+n_users, embedding_dim]
        """
        user_embeddings = self.user_embedding.weight
        item_embeddings = self.item_embedding.weight
        ego_embeddings = torch.cat([user_embeddings, item_embeddings], dim=0)
        return ego_embeddings

    def forward(self):
        all_embeddings = self.get_ego_embeddings()
        embeddings_list = [all_embeddings]

        for layer_idx in range(self.n_layers):
            all_embeddings = torch.sparse.mm(self.norm_adj_matrix, all_embeddings)
            embeddings_list.append(all_embeddings)
        lightgcn_all_embeddings = torch.stack(embeddings_list, dim=1)
        lightgcn_all_embeddings = torch.mean(lightgcn_all_embeddings, dim=1)

        user_all_embeddings, item_all_embeddings = torch.split(lightgcn_all_embeddings, [self.n_users, self.n_items])
        return user_all_embeddings, item_all_embeddings

    def calculate_loss(self, interaction, epoch_idx):
        # clear the storage variable when training
        if self.restore_user_e is not None or self.restore_item_e is not None:
            self.restore_user_e, self.restore_item_e = None, None

        user = interaction[self.USER_ID]
        pos_item = interaction[self.ITEM_ID]
        neg_item = interaction[self.NEG_ITEM_ID]

        user_all_embeddings, item_all_embeddings = self.forward()
        u_embeddings = user_all_embeddings[user]
        pos_embeddings = item_all_embeddings[pos_item]
        neg_embeddings = item_all_embeddings[neg_item]

        # calculate BPR Loss
        pos_scores = torch.mul(u_embeddings, pos_embeddings).sum(dim=1)
        neg_scores = torch.mul(u_embeddings, neg_embeddings).sum(dim=1)
        mf_loss = self.mf_loss(pos_scores, neg_scores)

        # calculate BPR Loss
        u_ego_embeddings = self.user_embedding(user)
        pos_ego_embeddings = self.item_embedding(pos_item)
        neg_ego_embeddings = self.item_embedding(neg_item)

        reg_loss = self.reg_loss(u_ego_embeddings, pos_ego_embeddings, neg_ego_embeddings)
        
        # reg_loss = reg_loss / len(interaction)

        loss = mf_loss + self.reg_weight * reg_loss

        return loss

    def predict(self, interaction):
        user = interaction[self.USER_ID]
        item = interaction[self.ITEM_ID]

        user_all_embeddings, item_all_embeddings = self.forward()

        u_embeddings = user_all_embeddings[user]
        i_embeddings = item_all_embeddings[item]
        scores = torch.mul(u_embeddings, i_embeddings).sum(dim=1)
        return scores

    def full_sort_predict(self, interaction):
        user = interaction[self.USER_ID]
        if self.restore_user_e is None or self.restore_item_e is None:
            self.restore_user_e, self.restore_item_e = self.forward()
        # get user embedding from storage variable
        u_embeddings = self.restore_user_e[user]

        # dot with all item embedding to accelerate
        scores = torch.matmul(u_embeddings, self.restore_item_e.transpose(0, 1))

        return scores.view(-1)

```

**Comment by h12345jack** [comment]
@Wicknight 我是非常愿意来提交pull request，但是```EmbLoss```和```RegLoss```被广泛使用。牵一发动全身...

因此,我非常建议构建一个bench mark的配置结果，让大家很容易的复现论文的结果。

**Comment by A11en0** [comment]
给大佬们点赞！
